_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+  primitive set of operations than ordinary Rust. This simplification
+  enables us to do a lot of things that were forbodingly complex
+  before. We'll look at one such case in those post -- non-zeroing
+  drop -- but as we'll see at the end, there are already many others
+  in the pipeline.
+
+Needless to say, we're excited, and the Rust community has stepped up
+in a big way to make MIR a reality. The compiler can bootstrap and run
+its test suite using MIR, and these tests have to pass on every new
+commit. Once we're able to run [Crater][crater] with MIR enabled and
+see no regressions across the entire [crates.io](http://crates.io/)
+ecosystem, we'll turn it on by default (or, you'll forgive a terrible
+(wonderful) pun, [launch MIR into orbit][orbit]).
+
+This blog post begins with an overview of MIR's design, demonstrating
+some of the ways that MIR is able to abstract away the full details of
+the Rust language. Next, we look at how MIR will help with
+implementing [non-zeroing drops][rfc320], a long-desired
+optimization. If after this post you find you are hungry for more,
+have a look at the [RFC that introduced MIR][rfc1211], or jump right
+into [the code][].
+
+<!-- more -->
+
+### Reducing Rust to a simple core
+
+MIR reduces Rust down to a simple core, removing almost all of the
+Rust syntax that you use every day, such as `for` loops, `match`
+expressions, and even method calls. Instead, those constructs are
+translated to a small set of primitives.
+
+The simple core language that MIR supports is not something you would
+want to program in. As we'll see, it makes things almost painfully
+explicit. But it's great if you want to write a type-checker or
+generate assembly code, as you now only have to handle the core
+operations that remain after MIR translation.
+
+To see what I mean, let's consider how we might simplify this fragment
+of code, which iterates over all the elements in a vector and
+processes them one by one:
+
+```rust
+for elem in vec {
+    process(elem);
+}
+```
+
+Rust itself offers three kinds of loops: `for` loops, like this one;
+`while` and `while let` loops, that iterate until some condition is
+met; and finally the simple `loop`, which just iterates until you
+break out of it. Each of these kinds of loops encapsulates a
+particular pattern, so they are quite useful when writing code. But
+for MIR, we'd like to reduce all of these into one core concept.
+
+A `for` loop in Rust works by converting a value into an iterator and
+then repeatedly calling `next` on that iterator. That means that we
+can rewrite the `for` loop we saw before into a `while let` loop
+that looks like this:
+
+```rust
+let mut iterator = vec.into_iter();
+while let Some(elem) = iterator.next() {
+    process(elem);
+}
+```
+
+By applying this rewriting, we can remove all `for` loops, but that
+still leaves multiple kinds of loops. So next we can imagine rewrite
+all `while let` loops into a simple `loop` combined with a `match`:
```

Comment:
Maybe worth mentioning that this step is done in the AST-> HIR lowering (you already mentioned the HIR in the intro)